src/main/scala/org/apache/spark/shuffle/ucx/UcxShuffleTransport.scala

```diff
@@ -0,0 +1,229 @@
+      val result = new UcxWorkerWrapper(localWorker, this, ucxShuffleConf, memoryPool)
+      allocatedWorkers.add(result)
+      result
+    })
+
+    progressThread.start()
+    result
+  }
+
+  /**
+   * Close all transport resources
+   */
+  override def close(): Unit = {
+    progressThread.interrupt()
+    globalWorker.signal()
+    try {
+      progressThread.join()
+    } catch {
+      case _:InterruptedException =>
+      case e:Throwable => logWarning(e.getLocalizedMessage)
+    }
+
+    memoryPool.close()
+    clientConnections.forEachValue(1, ep => ep.close())
+    registeredBlocks.forEachKey(1, blokId => unregister(blokId))
+    allocatedWorkers.forEach(_.close())
+    globalWorker.close()
+    ucxContext.close()
+  }
+
+  /**
+   * Add executor's worker address. For standalone testing purpose and for implementations that makes
+   * connection establishment outside of UcxShuffleManager.
+   */
+  def addExecutor(executorId: String, workerAddress: ByteBuffer): Unit = {
+    executorIdToAddress.put(executorId, workerAddress)
+  }
+
+  /**
+   * On a sender side process request of fetchBlocksByBlockIds
+   */
+  private[ucx] def processFetchBlocksRequest(memoryBlock: MemoryBlock) {
+    val buffer = UcxUtils.getByteBufferView(memoryBlock.address, memoryBlock.size.toInt)
+    val tag = buffer.getLong()
+    val workerAddressSize = buffer.getInt()
+    val workerAddress = buffer.slice()
+
+    workerAddress.limit(workerAddressSize)
+    buffer.position(buffer.position() + workerAddressSize)
+
+    val ep = clientConnections.computeIfAbsent(workerAddress, address => {
+      globalWorker.newEndpoint(new UcpEndpointParams().setUcpAddress(address))
+    })
+
+    val blockId: BlockId =
+      Utils.tryWithResource(new ByteBufferBackedInputStream(buffer))(bin => {
+        val in = new ObjectInputStream(bin)
+        in.readObject().asInstanceOf[BlockId]
+    })
+
+    val blockMemory = registeredBlocks.get(blockId).getMemoryBlock
+
+    ep.sendTaggedNonBlocking(blockMemory.address, blockMemory.size, tag, null)
+  }
+
+  private[ucx] def pinMemory(block: Block): UcpMemory = {
+    val blockMemory = block.getMemoryBlock
+    ucxContext.memoryMap(
+      memMapParams.setAddress(blockMemory.address).setLength(blockMemory.size))
+  }
+
+  /**
+   * Registers blocks using blockId on SERVER side.
+   */
+  override def register(blockId: BlockId, block: Block) {
+    val registeredBock: Block = if (ucxShuffleConf.pinMemory) {
+      logInfo(s"Pinning memory")
+      UcxPinnedBlock(block, pinMemory(block))
+    } else {
+      block
+    }
+    registeredBlocks.put(blockId, registeredBock)
+  }
+
+  /**
+   * Change location of underlying blockId in memory
+   */
+  override def mutate(blockId: BlockId, block: Block, callback: OperationCallback): Unit = {
+
+  }
+
+  /**
+   * Indicate that this blockId is not needed any more by an application
+   */
+  override def unregister(blockId: BlockId): Unit = {
```

Comment:
I thought unregister would need to check if the block is being served? e.g. that there may be some locking involved. Maybe this impl is not fully plumbed?